metric_learn/scml.py

```diff
@@ -0,0 +1,627 @@
+    self.beta = beta
+    self.basis = basis
+    self.n_basis = n_basis
+    self.gamma = gamma
+    self.max_iter = max_iter
+    self.output_iter = output_iter
+    self.verbose = verbose
+    self.preprocessor = preprocessor
+    self.random_state = random_state
+    super(_BaseSCML, self).__init__(preprocessor)
+
+  def _fit(self, triplets, basis=None, n_basis=None):
+    """
+    Optimization procedure to find a sparse vector of weights to
+    construct the metric from the basis set. This is based on the
+    dual averaging method.
+    """
+
+    # Currently prepare_inputs makes triplets contain points and not indices
+    triplets = self._prepare_inputs(triplets, type_of_inputs='tuples')
+
+    # TODO:
+    # This algorithm is built to work with indices, but in order to be
+    # compliant with the current handling of inputs it is converted
+    # back to indices by the following function. This should be improved
+    # in the future.
+    triplets, X = self._to_index_points(triplets)
+
+    if basis is None:
+      basis, n_basis = self._initialize_basis(triplets, X)
+
+    dist_diff = self._compute_dist_diff(triplets, X, basis)
+
+    n_triplets = triplets.shape[0]
+
+    w = np.zeros((1, n_basis))
+    avg_grad_w = np.zeros((1, n_basis))
+
+    best_obj = np.inf
+
+    rng = check_random_state(self.random_state)
+    rand_int = rng.randint(low=0, high=n_triplets, size=self.max_iter)
+    for iter in range(self.max_iter):
+      if iter % self.output_iter == 0:
+        # regularization part of obj function
+        obj1 = np.sum(w)*self.beta
+
+      # Every triplet distance difference in the space given by L
+      # plus a slack of one
+        slack_val = 1 + np.matmul(dist_diff, w.T)
+      # Mask of places with positive slack
+        slack_mask = slack_val > 0
+
+        # loss function of learning task part of obj function
+        obj2 = sum_where(slack_val, slack_mask)/n_triplets
+
+        obj = obj1 + obj2
+        if self.verbose:
+          count = np.sum(slack_mask)
+          print("[Global] iter %d\t obj %.6f\t num_imp %d" % (iter,
+                obj, count))
+
+        # update the best
+        if obj < best_obj:
+          best_obj = obj
+          best_w = w
+
+      # TODO:
+      # Maybe allow the usage of mini-batch opt?
+
+      idx = rand_int[iter]
+
+      slack_val = 1 + np.matmul(dist_diff[idx, :], w.T)
+
+      if slack_val > 0:
+        avg_grad_w = (iter * avg_grad_w + dist_diff[idx, :]) / (iter+1)
+      else:
+        avg_grad_w = iter * avg_grad_w / (iter+1)
+
+      scale_f = -np.sqrt(iter+1) / self.gamma
+
+      # proximal operator with negative trimming equivalent
+      w = scale_f * np.minimum(avg_grad_w + self.beta, 0)
+
+    if self.verbose:
+      print("max iteration reached.")
+
+    # return L matrix yielded from best weights
+    self.components_ = self._components_from_basis_weights(basis, best_w)
+
+    return self
+
+  def _compute_dist_diff(self, T, X, basis):
```

Comment:
Maybe rename "T" to "triplets" or "trip" or something like this ? (so as not to have the same name as the transpose operator)